### PR TITLE
Automatic identification of Boundaries based on continuity

### PR DESCRIPTION
**Pull request by jzwar**
# Overview
The idea is simple, if two adjacent elements share a boundary that connect G1-continouous, they probably belong to the same boundary on a global scale. This hopefully facilitates the application of boundary conditions in IGA.

## Addressed issues
- #81 

## Showcase
TBD

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [x] Added test(s)

### CODE REVIEW

**Review by j042 — COMMENTED** [review]
Good stuff

File: cpp/splinepy/py/py_spline_exporter.hpp [hunk 1]
```diff
@@ -1157,4 +1157,256 @@ py::tuple RetrieveMfemInformation(const py::array_t<double>& py_corner_vertices,
...
+  const int* boundary_interfaces_ptr =
+      static_cast<int*>(boundary_interfaces.request().ptr);
+  int* global_interfaces_ptr =
+      static_cast<int*>(global_interfaces.request().ptr);
+
+  // Auxiliary Lambdas to keep code clean
+  // Check if to tangential vectors are g1 (tol > cos(phi))
+  auto areG1 = [&tolerance, &dim_](const std::vector<double>& vec0,
+                                   const std::vector<double>& vec1) -> bool {
+    // Checks in Debug
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_spline_exporter.hpp:1245]: Done

File: splinepy/multipatch.py [hunk 2]
```diff
@@ -154,6 +173,75 @@ def boundaries(self):
...
+    def boundary_patches(self):
+        """Extract all boundary patches of a given Multipatch system as splines
+
+        Returns
+        -------
+        boundary_patches : list
+          list of all embedded splines
+        """
+        from splinepy import settings
+        from splinepy.splinepy_core import extract_all_boundary_splines
```
[the remark this answers is not in the document]
Comment (jzwar) [reply splinepy/multipatch.py]: Done

File: splinepy/multipatch.py [hunk 3]
```diff
@@ -154,6 +173,75 @@ def boundaries(self):
...
+        Returns
+        -------
+        np.ndarray : connectivity
+          Connectivity in local patch enumeration system
+        """
+        from splinepy import settings
+        from splinepy.splinepy_core import (
+            boundary_centers,
+            interfaces_from_boundary_centers,
+        )
```
[the remark this answers is not in the document]
Comment (jzwar) [reply splinepy/multipatch.py]: Done

File: splinepy/multipatch.py [hunk 4]
```diff
@@ -154,6 +173,75 @@ def boundaries(self):
...
+            patches = extract_all_boundary_splines(
+                self.splines, self.interfaces
+            )
+            self._boundary_splines = [
+                settings.NAME_TO_TYPE[p.name](spline=p) for p in patches
+            ]
+
+        return self._boundary_splines
+
+    @property
```
[the remark this answers is not in the document]
Comment (jzwar) [reply splinepy/multipatch.py:203]: I wanted to keep the interfaces as easy as possible. So making this a property was intentional. In the interface property it gets computed automatically if it is not yet provided. Here, it makes no sense to provide the interface boundaries, as they are a function of the interfaces. 

Can we maybe push that to the next pr?

File: splinepy/multipatch.py [hunk 5]
```diff
@@ -154,6 +173,75 @@ def boundaries(self):
...
+            # Get Boundary Centers
+            boundary_patch_face_centers = np.vstack(
+                [boundary_centers(s) for s in self.boundary_patches]
+            )
+
+            # Using the setter instead of the the member, all necessery
+            # checks will be performed
+            self._boundary_interfaces = interfaces_from_boundary_centers(
+                boundary_patch_face_centers,
+                settings.TOLERANCE,
```
[the remark this answers is not in the document]
Comment (jzwar) [reply splinepy/multipatch.py:228]: I think there is no need to adapt the tolerances here, as the boundary interfaces are only really required for the `boundary_from_continuity` function. All the cost intensive property calls are just needed there. So I think it is more important to keep things simple here

**Comment by jzwar** [comment]
I just realized, that my approach was kind of stupid. As boundary patches I could return a Multipatch object and reuse all the code :facepalm:

**Comment by jzwar** [comment]
image

It seems to work properly, although the plotting is still wip

File: cpp/splinepy/py/py_spline_exporter.hpp [hunk 6]
```diff
@@ -1157,4 +1157,256 @@ py::tuple RetrieveMfemInformation(const py::array_t<double>& py_corner_vertices,
...
+  }
+  // Auxiliary data
+  py::list boundary_splines{};
+  const int* interface_ptr = static_cast<int*>(interfaces.request().ptr);
+  const int n_patches = interfaces.shape(0);
+  const int n_faces = interfaces.shape(1);
+  const int para_dim_ = n_faces / 2;
+  const auto cpp_spline_list =
+      ListOfPySplinesToVectorOfCoreSplines(spline_list);
+  for (int i{}; i < n_patches; i++) {
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_spline_exporter.hpp:1186]: alright, I will implement some list-like container for multi-patch splines

**Review by j042 — CHANGES_REQUESTED** [review]

File: splinepy/multipatch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (j042) [reply splinepy/multipatch.py]: I don't think defining a function with `@property` make it necessarily easy. It can be an empty arg function call.

File: splinepy/multipatch.py [hunk 7]
```diff
@@ -154,6 +199,29 @@ def boundaries(self):
...
+    @property
+    def boundary_patches(self):
+        """Extract all boundary patches of a given Multipatch system as splines
+
+        Returns
+        -------
+        boundary_patches : list
+          list of all embedded splines
+        """
+        if self._boundary_splines is None:
```
Comment (j042) [inline splinepy/multipatch.py:211]: This doesn't work well with inplace changes

File: splinepy/multipatch.py [hunk 8]
```diff
@@ -154,6 +173,75 @@ def boundaries(self):
 
         return boundary_list
 
+    @property
```
[the remark this answers is not in the document]
Comment (j042) [reply splinepy/multipatch.py:202]: keep it as a normal function for the future?

File: splinepy/multipatch.py (hunk 7 above)
Comment (jzwar) [reply splinepy/multipatch.py:211]: This is true, I am not yet sure how to handle the in-place changes at all, especially if splines are modified on the fly. The only safeguard I have at this point is in the splines-setter, where - if called - all dependent members are reinitialized... I have no better solution at this point

File: CMakeLists.txt [hunk 9]
```diff
@@ -63,6 +63,10 @@ if(SPLINEPY_MORE)
     set(SPLINEPY_DEFS ${SPLINEPY_DEFS} SPLINEPY_MORE)
 endif(SPLINEPY_MORE)
 
+if ($<NOT:$<CONFIG:Debug>)
+    set(SPLINEPY_DEFS ${SPLINEPY_DEFS} NDEBUG)
+endif()
+
```
[the remark this answers is not in the document]
Comment (jzwar) [reply CMakeLists.txt]: Done

File: cpp/splinepy/py/py_spline_exporter.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_spline_exporter.hpp]: I think I found a solution to this, check out the new exporter

File: CMakeLists.txt [hunk 10]
```diff
@@ -30,6 +30,7 @@ set(namespace "${PROJECT_NAME}::")
 set(CMAKE_CXX_VISIBILITY_PRESET "hidden")
 set(SPLINEPY_FLAGS -fPIC)
 set(SPLINEPY_OPTIMIZATION_FLAGS $<$<NOT:$<CONFIG:Debug>>:-O3>)
+set(SPLINEPY_DEFS ${SPLINEPY_DEFS} $<$<NOT:$<CONFIG:Debug>>:NDEBUG>)
```
Comment (j042) [inline CMakeLists.txt:33]:
```suggestion
set(SPLINEPY_DEFS $<$<NOT:$<CONFIG:Debug>>:NDEBUG>)
```

File: splinepy/multipatch.py [hunk 11]
```diff
@@ -313,14 +374,18 @@ def add_boundary_with_function(
...
+            boundary_ids = np.isin(-self.interfaces, np.abs(from_boundaries))
 
         # Check if there is a boundary
         if not boundary_ids.any():
-            raise ValueError("No boundary elements could be identified")
+            self.logd(
+                "No boundary elements could be identified that match "
+                "requirements"
+            )
+            return
```
Comment (j042) [inline splinepy/multipatch.py:388]:
```suggestion
            return None
```

File: splinepy/multipatch.py [hunk 12]
```diff
@@ -344,21 +409,55 @@ def add_boundary_with_function(
...
+        self._logd("Start propagation of information...")
+        n_new_boundaries = boundaries_from_continuity(
+            self.boundary_patches().splines,
+            self.boundary_patches().interfaces,
+            self.interfaces,
+            settings.TOLERANCE,
+            settings.NTHREADS,
+        )
+        self._logd(f"{n_new_boundaries} new boundaries were assigned")
+        return
```
Comment (j042) [inline splinepy/multipatch.py:437]:
```suggestion
        return None
```

File: splinepy/multipatch.py [hunk 13]
```diff
@@ -344,21 +409,55 @@ def add_boundary_with_function(
...
+        """
+        # retrieve all boundary elements
+        boundary_ids = np.isin(-self.interfaces, np.abs(bid_list))
+
+        if not boundary_ids.any():
+            self.logd(
+                "No boundary elements could be identified that match "
+                "requirements"
+            )
+            return
```
Comment (j042) [inline splinepy/multipatch.py:461]:
```suggestion
            return None
```

File: splinepy/multipatch.py [hunk 14]
```diff
@@ -344,21 +409,55 @@ def add_boundary_with_function(
...
 
-        pass
+        # Pass information to c++ backend
+        self._logd("Start propagation of information...")
+        n_new_boundaries = boundaries_from_continuity(
+            self.boundary_patches().splines,
+            self.boundary_patches().interfaces,
+            self.interfaces,
+            settings.TOLERANCE,
+            settings.NTHREADS,
```
Comment (j042) [inline splinepy/multipatch.py:434]: make this function param?

File: splinepy/multipatch.py [hunk 15]
```diff
@@ -270,10 +332,10 @@ def determine_interfaces(self):
 
         return self.interfaces
 
-    def add_boundary_with_function(
+    def boundary_from_function(
```
Comment (j042) [inline splinepy/multipatch.py:335]:
```suggestion
    def boundaries_from_function(
```
or you can change all the other one to `boundary_*`

File: splinepy/multipatch.py [hunk 16]
```diff
@@ -270,10 +332,10 @@ def determine_interfaces(self):
...
         self,
         function,
-        only_unassigned=False,
+        from_boundaries=None,
```
Comment (j042) [inline splinepy/multipatch.py:338]:
```suggestion
        mask=None,
```
boundaries from function from boundaries is a bit unclear

File: splinepy/multipatch.py (hunk 7 above)
Comment (j042) [reply splinepy/multipatch.py:209]: I'd say, we can leave this to always recompute or add a `use_saved` param until we can handle in-place changes

**Review by j042 — COMMENTED** [review]
Some of the docstrings are incomplete (splines, interfaces, ....)
Other than that, thanks for updating!

File: splinepy/multipatch.py (hunk 15 above)
Comment (jzwar) [reply splinepy/multipatch.py]: The others auto assign multiple boundaries, this one does one at a time. I reconsidered this change and changed it back, :sweat_smile:

File: splinepy/multipatch.py [hunk 17]
```diff
@@ -313,14 +416,18 @@ def add_boundary_with_function(
...
+        if mask is None:
             boundary_ids = self.interfaces < 0
         else:
-            boundary_ids = self.interfaces == -1
+            boundary_ids = np.isin(-self.interfaces, np.abs(mask))
 
         # Check if there is a boundary
         if not boundary_ids.any():
-            raise ValueError("No boundary elements could be identified")
+            self.logd(
```
Comment (j042) [inline splinepy/multipatch.py:426]:
```suggestion
            self._logd(
```

File: splinepy/multipatch.py [hunk 18]
```diff
@@ -344,21 +451,72 @@ def add_boundary_with_function(
...
+
+        Returns
+        -------
+        None
+        """
+        # retrieve all boundary elements
+        boundary_ids = np.isin(-self.interfaces, np.abs(mask))
+
+        if not boundary_ids.any():
+            self.logd(
```
Comment (j042) [inline splinepy/multipatch.py:516]:
```suggestion
            self._logd(
```

**Review by j042 — APPROVED** [review]
Thanks for this new feature :) 
Please update those comments before merging
